### PR TITLE
Unlabelled dataset support WIP

### PR DESCRIPTION
**Pull request by proteneer**
@rbharath 

This is a WIP PR for #375 - I wanted to get your thoughts to make sure I'm along the right track before I start adding in tests to this. This ended up being a little more involved that I had initially thought:

1. I wanted to make sure get_shape() works correctly when labels are omitted, the PR changes it so that it turns empty tuples for the y and w. 
2. Only the appropriate joblib files are written out. ie. If it's missing y values then joblib-y isn't written.
3. This changes the order of the columns from 

```
columns = ('basename', 'ids', 'X', 'task_names', 'y', 'w')
```
to
```
columns = ('basename','task_names', 'ids', 'X', 'y', 'w'))
```
for compatibility reasons.

### CODE REVIEW

File: deepchem/data/datasets.py [hunk 1]
```diff
@@ -526,15 +539,22 @@ def iterate(dataset):
...
-        y = np.array(load_from_disk(
-            os.path.join(dataset.data_dir, row['y'])))
-        w_filename = os.path.join(dataset.data_dir, row['w'])
-        if os.path.exists(w_filename):
-            w = np.array(load_from_disk(w_filename))
-        else:
-            w = np.ones(y.shape)
         ids = np.array(load_from_disk(
             os.path.join(dataset.data_dir, row['ids'])), dtype=object)
+        # These columns may be missing is the dataset is unlabelled.
```
Comment (lilleswing) [inline deepchem/data/datasets.py]: These columns may be missing *if* the dataset is unlabelled

Comment (proteneer) [inline deepchem/data/datasets.py]: Good catch - I'll be sure to fix all of the typos in the final pass

**Comment by proteneer** [comment]
It also looks like getting this to work for the LSTM/one-shot approach will be a bit of a pain (especially since predict_proba_on_batch assumes that the test set contains labelled y's when constructing the feed dict)

```  python
# support_classifier.py
  def construct_feed_dict(self, test, support, training=True, add_phase=False):
    """Constructs tensorflow feed from test/support sets."""
    # Generate dictionary elements for support 
    feed_dict = (
        self.model.support_graph_topology.batch_to_feed_dict(support.X))
    feed_dict[self.support_label_placeholder] = np.squeeze(support.y)
    # Get graph information for test 
    batch_topo_dict = (
        self.model.test_graph_topology.batch_to_feed_dict(test.X))
    feed_dict = merge_dicts([batch_topo_dict, feed_dict])
    # Generate dictionary elements for test
    feed_dict[self.test_label_placeholder] = np.squeeze(test.y)
    feed_dict[self.test_weight_placeholder] = np.squeeze(test.w)

    if add_phase:
      feed_dict[self.phase] = training
    return feed_dict

**Comment by rbharath** [comment]
@proteneer Thanks for taking this on. It's not an easy task to start digging around in the DeepChem backends. Documentation is quite sparse for non-user-exposed parts of the code. Here are some first thoughts
- We probably can't change the order of the columns. This would be a major breaking change, since it would require us to regenerate many existing, processed datasets (correct me if I'm wrong here).
- For one-shot, the test-label-placeholder isn't used during evaluation. I believe it should still run if you don't bother setting the placeholder. There might be a little backend clean-up required to make this work out though, since it might be computing (but not returning) the loss even during eval.

**Comment by proteneer** [comment]
OK - we can probably get away with not having to change the order of columns. Though this also means that presumably there's an implicit versioning semantic now associated with these data files (which will sooner or later need to be correctly implemented)

**Comment by rbharath** [comment]
Yes, you're absolutely right that the datasets objects are version dependent. It's worth getting the versioning semantic right. Would love to hear some thoughts on #376 about how we should handle versioning in the alpha release.

**Comment by proteneer** [comment]
I'm also slightly at a loss about what to do with BalancingTransformers (which create initial weights based on the population of negs and positives from the labels) using during training. With true prospective data, I obviously can't do this - do I just not give it a transformer when running predict_proba (since it takes in a transformer parameter) ?

**Comment by rbharath** [comment]
The easiest way is probably just to have `BalancingTransformer` do nothing if weights/labels aren't available (just return the dataset unchanged). This would be a non-damaging change since it wouldn't break any existing pipelines.

**Comment by proteneer** [comment]
Are subsequent commits to a PR not ran automatically on TravisCI? is there a way to trigger them?

**Comment by rbharath** [comment]
You might perhaps not be at head? (I think a couple small PRs were merged in yesterday). Try running `git pull origin master` and pushing to this PR again.

**Comment by proteneer** [comment]
I just did a rebase.

**Comment by rbharath** [comment]
Maybe just push a dummy commit then. Sometimes, Travis drops the ball at random.

**Comment by proteneer** [comment]
Seems still to be stuck - I don't have write access to this repo so I can't force a rebuild manually.

**Comment by rbharath** [comment]
Seems to have triggered now https://travis-ci.org/deepchem/deepchem/builds/197767376

**Comment by proteneer** [comment]
Added bugfixes and additional tests for transformers on unlabeled data. I think the behavior is most sane if we simply raise an exception when asked to transform unlabelled datasets on the y and w axes, though I'm open to suggestions.

**Comment by rbharath** [comment]
@proteneer Design looks reasonable at first pass :-). Since this is a change to the backend, would you mind running all the benchmark models on the Tox21 collection and pasting the output below? (Running `python benchmark.py -h` from within the examples/ folder will give directions on how to do this). Will do a careful review once we verify that Tox21 benchmark numbers look sane.

**Comment by proteneer** [comment]
@rbharath Do we need to support the NumpyDataset class? I noticed that the CSVLoader hardcode in the DiskDataset class - I wasn't sure about the other Dataset types.

**Comment by proteneer** [comment]
@rbharath full output

```
(venv) yutong@symmetry:~/Code/deepchem/examples$ python benchmark.py -m graphconv -d tox21
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcurand.so locally
Using TensorFlow backend.
-------------------------------------
Benchmark graphconv on dataset: tox21
-------------------------------------
Splitting function: index
About to featurize Tox21 dataset.
Loading raw samples now.
shard_size: 8192
About to start loading CSV from /home/yutong/Code/deepchem/examples/tox21/../../datasets/tox21.csv.gz
Loading shard 1 of size 8192.
Featurizing sample 0
Featurizing sample 1000
Featurizing sample 2000
Featurizing sample 3000
Featurizing sample 4000
Featurizing sample 5000
Featurizing sample 6000
Featurizing sample 7000
Featurizing sample 8000
TIMING: featurizing shard 0 took 12.521 s
TIMING: dataset construction took 14.930 s True
Loading dataset from disk.
About to transform data
TIMING: dataset construction took 3.326 s True
Loading dataset from disk.
TIMING: dataset construction took 2.978 s True
Loading dataset from disk.
TIMING: dataset construction took 1.268 s True
Loading dataset from disk.
TIMING: dataset construction took 1.330 s True
Loading dataset from disk.
I tensorflow/core/common_runtime/gpu/gpu_device.cc:885] Found device 0 with properties: 
name: TITAN X (Pascal)
major: 6 minor: 1 memoryClockRate (GHz) 1.531
pciBusID 0000:03:00.0
Total memory: 11.90GiB
Free memory: 10.83GiB
I tensorflow/core/common_runtime/gpu/gpu_device.cc:906] DMA: 0 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:916] 0:   Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: TITAN X (Pascal), pci bus id: 0000:03:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: TITAN X (Pascal), pci bus id: 0000:03:00.0)
/home/yutong/venv/lib/python3.5/site-packages/tensorflow/python/ops/gradients_impl.py:91: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
-------------------------------------
Start fitting by graph convolution
Training for 15 epochs
Starting epoch 0
On batch 0
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 2267 get requests, put_count=1958 evicted_count=1000 eviction_rate=0.510725 and unsatisfied allocation rate=0.621526
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 100 to 110
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 2338 get requests, put_count=2377 evicted_count=1000 eviction_rate=0.420698 and unsatisfied allocation rate=0.420873
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 256 to 281
On batch 50
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 12845 get requests, put_count=12844 evicted_count=1000 eviction_rate=0.0778574 and unsatisfied allocation rate=0.0825224
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 655 to 720
On batch 100
Starting epoch 1
On batch 0
On batch 50
On batch 100
Starting epoch 2
On batch 0
On batch 50
On batch 100
Starting epoch 3
On batch 0
On batch 50
On batch 100
Starting epoch 4
On batch 0
On batch 50
On batch 100
Starting epoch 5
On batch 0
On batch 50
On batch 100
Starting epoch 6
On batch 0
On batch 50
On batch 100
Starting epoch 7
On batch 0
On batch 50
On batch 100
Starting epoch 8
On batch 0
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 159939 get requests, put_count=159946 evicted_count=1000 eviction_rate=0.00625211 and unsatisfied allocation rate=0.00717148
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 1694 to 1863
On batch 50
On batch 100
Starting epoch 9
On batch 0
On batch 50
On batch 100
Starting epoch 10
On batch 0
On batch 50
On batch 100
Starting epoch 11
On batch 0
On batch 50
On batch 100
Starting epoch 12
On batch 0
On batch 50
On batch 100
Starting epoch 13
On batch 0
On batch 50
On batch 100
Starting epoch 14
On batch 0
On batch 50
On batch 100
computed_metrics: [0.90003048691435927, 0.94671640981650751, 0.90880816665990505, 0.90054930841080338, 0.78350074430140881, 0.89629869951523644, 0.9299374944725658, 0.82646071792894604, 0.92837086933087964, 0.8673871069562531, 0.90807645419714389, 0.90406853940193299]
computed_metrics: [0.82623394185260302, 0.79847849712777519, 0.8585407647907648, 0.80701953657428438, 0.67450271247739602, 0.70445287259446543, 0.81920903954802249, 0.81505646768529938, 0.8405295919375555, 0.75164995822890557, 0.8578622482131254, 0.84749999999999992]
-------------------------------------
Benchmark graphconv on dataset: tox21
-------------------------------------
Splitting function: random
About to featurize Tox21 dataset.
Loading raw samples now.
shard_size: 8192
About to start loading CSV from /home/yutong/Code/deepchem/examples/tox21/../../datasets/tox21.csv.gz
Loading shard 1 of size 8192.
Featurizing sample 0
Featurizing sample 1000
Featurizing sample 2000
Featurizing sample 3000
Featurizing sample 4000
Featurizing sample 5000
Featurizing sample 6000
Featurizing sample 7000
Featurizing sample 8000
TIMING: featurizing shard 0 took 12.674 s
TIMING: dataset construction took 14.975 s True
Loading dataset from disk.
About to transform data
TIMING: dataset construction took 3.100 s True
Loading dataset from disk.
TIMING: dataset construction took 3.438 s True
Loading dataset from disk.
TIMING: dataset construction took 1.105 s True
Loading dataset from disk.
TIMING: dataset construction took 1.481 s True
Loading dataset from disk.
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: TITAN X (Pascal), pci bus id: 0000:03:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: TITAN X (Pascal), pci bus id: 0000:03:00.0)
/home/yutong/venv/lib/python3.5/site-packages/tensorflow/python/ops/gradients_impl.py:91: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
-------------------------------------
Start fitting by graph convolution
Training for 15 epochs
Starting epoch 0
On batch 0
On batch 50
On batch 100
Starting epoch 1
On batch 0
On batch 50
On batch 100
Starting epoch 2
On batch 0
On batch 50
On batch 100
Starting epoch 3
On batch 0
On batch 50
On batch 100
Starting epoch 4
On batch 0
On batch 50
On batch 100
Starting epoch 5
On batch 0
On batch 50
On batch 100
Starting epoch 6
On batch 0
On batch 50
On batch 100
Starting epoch 7
On batch 0
On batch 50
On batch 100
Starting epoch 8
On batch 0
On batch 50
On batch 100
Starting epoch 9
On batch 0
On batch 50
On batch 100
Starting epoch 10
On batch 0
On batch 50
On batch 100
Starting epoch 11
On batch 0
On batch 50
On batch 100
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 841805 get requests, put_count=841795 evicted_count=2000 eviction_rate=0.00237588 and unsatisfied allocation rate=0.00281775
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 3987 to 4385
Starting epoch 12
On batch 0
On batch 50
On batch 100
Starting epoch 13
On batch 0
On batch 50
On batch 100
Starting epoch 14
On batch 0
On batch 50
On batch 100
computed_metrics: [0.90532777371914552, 0.94882872363715975, 0.91216648932314492, 0.90358798540327934, 0.77462758263053022, 0.88245484176763145, 0.92530115309126604, 0.83528656573801929, 0.92475065156375302, 0.87266694592616945, 0.90930080603327879, 0.90613588110403398]
computed_metrics: [0.78449333800841514, 0.92992883397200277, 0.87850594095869439, 0.88886919532080821, 0.72956259426847669, 0.8212612130137904, 0.85101794340924775, 0.80240365626586918, 0.84526281785038537, 0.74396411092985315, 0.90415846456692917, 0.86695369847543757]
-------------------------------------
Benchmark graphconv on dataset: tox21
-------------------------------------
Splitting function: scaffold
About to featurize Tox21 dataset.
Loading raw samples now.
shard_size: 8192
About to start loading CSV from /home/yutong/Code/deepchem/examples/tox21/../../datasets/tox21.csv.gz
Loading shard 1 of size 8192.
Featurizing sample 0
Featurizing sample 1000
Featurizing sample 2000
Featurizing sample 3000
Featurizing sample 4000
Featurizing sample 5000
Featurizing sample 6000
Featurizing sample 7000
Featurizing sample 8000
TIMING: featurizing shard 0 took 12.661 s
TIMING: dataset construction took 15.357 s True
Loading dataset from disk.
About to transform data
TIMING: dataset construction took 3.472 s True
Loading dataset from disk.
TIMING: dataset construction took 2.906 s True
Loading dataset from disk.
TIMING: dataset construction took 1.528 s True
Loading dataset from disk.
TIMING: dataset construction took 1.518 s True
Loading dataset from disk.
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: TITAN X (Pascal), pci bus id: 0000:03:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: TITAN X (Pascal), pci bus id: 0000:03:00.0)
/home/yutong/venv/lib/python3.5/site-packages/tensorflow/python/ops/gradients_impl.py:91: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
-------------------------------------
Start fitting by graph convolution
Training for 15 epochs
Starting epoch 0
On batch 0
On batch 50
On batch 100
Starting epoch 1
On batch 0
On batch 50
On batch 100
Starting epoch 2
On batch 0
On batch 50
On batch 100
Starting epoch 3
On batch 0
On batch 50
On batch 100
Starting epoch 4
On batch 0
On batch 50
On batch 100
Starting epoch 5
On batch 0
On batch 50
On batch 100
Starting epoch 6
On batch 0
On batch 50
On batch 100
Starting epoch 7
On batch 0
On batch 50
On batch 100
Starting epoch 8
On batch 0
On batch 50
On batch 100
Starting epoch 9
On batch 0
On batch 50
On batch 100
Starting epoch 10
On batch 0
On batch 50
On batch 100
Starting epoch 11
On batch 0
On batch 50
On batch 100
Starting epoch 12
On batch 0
On batch 50
On batch 100
Starting epoch 13
On batch 0
On batch 50
On batch 100
Starting epoch 14
On batch 0
On batch 50
On batch 100
computed_metrics: [0.89698201388389931, 0.96112911060492445, 0.91761118706834188, 0.92750419362568892, 0.79275359343682239, 0.90439634703196359, 0.94956898255600541, 0.84906028988529858, 0.94940385257925552, 0.88298578072021117, 0.92519913117222286, 0.92798012265583418]
computed_metrics: [0.75200278648554497, 0.85876716247139595, 0.8301879232279521, 0.74437910757523351, 0.69227348817744305, 0.71064382139148496, 0.67125835512932286, 0.72383526383526386, 0.72360465116279071, 0.76481408487715141, 0.80116496476341148, 0.77126047423783206]
```

**Comment by joegomes** [comment]
@proteneer Thanks for the PR!  I also think the point you raise about `NumpyDataset` support for other `DataLoader` classes is a good one, can you open an issue for discussion?  There is still the typo that @lilleswing found.

@rbharath I've compared Tox21 output and done a careful review of the code and LGTM after typo fix.

**Comment by proteneer** [comment]
Okay I'll do add in the changes and do a rebase on Monday. I think we can agree on deferring the NumpyDataset and Dataloader to a separate PR?

**Comment by rbharath** [comment]
@joegomes Thanks for reviewing!

@proteneer Since @joegomes has LGTM'ed (except the small typo), I'll go ahead and merge this one in.  I agree that the `NumpyDataset` and `DataLoader` fix can likely be dealt with in separate PR.

Congrats on your first major DeepChem backend change! Hope to see more coming our way :-)

**Pull request merged**